File: src/calibration/calibrator.py
"""Interactive checkerboard calibration workflow."""

from __future__ import annotations

from pathlib import Path

import cv2
import numpy as np

from src.calibration.checkerboard import (
    build_object_points,
    draw_checkerboard_overlay,
    find_checkerboard_corners,
)
from src.calibration.matrices import save_calibration_matrices
from src.config import load_section, resolve_config_path
from src.exceptions import CalibrationError


class CameraCalibrator:
    """Collects checkerboard samples and computes camera intrinsics."""

    def __init__(self, config_path: str | Path | None = None) -> None:
        self._config_path = resolve_config_path(config_path)
        cal_cfg = load_section("calibration", self._config_path)
        board_cfg = cal_cfg.get("checkerboard", {})

        self.cols = int(board_cfg.get("cols", 9))
        self.rows = int(board_cfg.get("rows", 6))
        self.square_size_mm = float(board_cfg.get("square_size_mm", 25.0))
        self.min_samples = int(cal_cfg.get("min_samples", 15))

        self._pattern_size = (self.cols, self.rows)
        self._object_template = build_object_points(self.cols, self.rows, self.square_size_mm)
        self._object_points: list[np.ndarray] = []
        self._image_points: list[np.ndarray] = []
        self._image_size: tuple[int, int] | None = None

        self.camera_matrix: np.ndarray | None = None
        self.dist_coeffs: np.ndarray | None = None
        self.reprojection_error: float | None = None

    @property
    def sample_count(self) -> int:
        return len(self._object_points)

    @property
    def is_calibrated(self) -> bool:
        return self.camera_matrix is not None and self.dist_coeffs is not None

    def find_checkerboard(
        self,
        frame: np.ndarray,
        *,
        robust: bool = False,
    ) -> tuple[bool, np.ndarray | None]:
        return find_checkerboard_corners(frame, self._pattern_size, robust=robust)

    def add_sample(self, frame: np.ndarray) -> bool:
        found, corners = self.find_checkerboard(frame, robust=True)
        if not found or corners is None:
            return False

        height, width = frame.shape[:2]
        if self._image_size is None:
            self._image_size = (width, height)
        elif self._image_size != (width, height):
            raise CalibrationError(
                "All calibration frames must have the same resolution. "
                f"Expected {self._image_size}, got {(width, height)}."
            )

        self._object_points.append(self._object_template.copy())
        self._image_points.append(corners)
        return True

    def calibrate(self) -> tuple[np.ndarray, np.ndarray, float]:
        if self.sample_count < self.min_samples:
            raise CalibrationError(
                f"Need at least {self.min_samples} valid samples, have {self.sample_count}."
            )
        if self._image_size is None:
            raise CalibrationError("No calibration samples collected.")

        error, camera_matrix, dist_coeffs, _rvecs, _tvecs = cv2.calibrateCamera(
            self._object_points,
            self._image_points,
            self._image_size,
            None,
            None,
        )

        self.camera_matrix = camera_matrix
        self.dist_coeffs = dist_coeffs
        self.reprojection_error = float(error)
        return camera_matrix, dist_coeffs, float(error)

    def draw_overlay(self, frame: np.ndarray) -> np.ndarray:
        found, corners = self.find_checkerboard(frame, robust=False)
        if found and corners is not None:
            overlay = draw_checkerboard_overlay(frame, self._pattern_size, corners)
        else:
            overlay = frame.copy()

        ready = self.sample_count >= self.min_samples
        status = (
            f"Samples: {self.sample_count}/{self.min_samples}  "
            f"{'READY' if ready else 'collect more'}"
        )
        cv2.putText(
            overlay,
            status,
            (20, 40),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.8,
            (0, 255, 0),
            2,
            cv2.LINE_AA,
        )
        return overlay

    def save_to_config(self, config_path: str | Path | None = None) -> Path:
        if not self.is_calibrated or self._image_size is None:
            raise CalibrationError("Calibrate before saving results.")

        assert self.camera_matrix is not None
        assert self.dist_coeffs is not None
        assert self.reprojection_error is not None

        return save_calibration_matrices(
            self.camera_matrix,
            self.dist_coeffs,
            self.reprojection_error,
            self._image_size,
            config_path or self._config_path,
        )

    def reset_samples(self) -> None:
        self._object_points.clear()
        self._image_points.clear()
        self._image_size = None

File: src/calibration/calibrator.cpp
#include "chess_robot/calibration/calibrator.hpp"

#include <sstream>

#include <opencv2/calib3d.hpp>
#include <opencv2/imgproc.hpp>

#include "chess_robot/calibration/checkerboard.hpp"
#include "chess_robot/calibration/matrices.hpp"
#include "chess_robot/config.hpp"
#include "chess_robot/exceptions.hpp"

namespace chess_robot {

CameraCalibrator::CameraCalibrator(const std::optional<fs::path>& config_path)
    : config_path_(resolveConfigPath(config_path)) {
  YAML::Node cal = loadSection("calibration", config_path_);
  YAML::Node board = cal["checkerboard"] ? cal["checkerboard"] : YAML::Node(YAML::NodeType::Map);
  cols_ = board["cols"] ? board["cols"].as<int>() : 9;
  rows_ = board["rows"] ? board["rows"].as<int>() : 6;
  square_size_mm_ = board["square_size_mm"] ? board["square_size_mm"].as<float>() : 25.0f;
  min_samples_ = cal["min_samples"] ? cal["min_samples"].as<int>() : 15;
  pattern_size_ = cv::Size(cols_, rows_);
  object_template_ = buildObjectPoints(cols_, rows_, square_size_mm_);
}

std::optional<cv::Mat> CameraCalibrator::findCheckerboard(const cv::Mat& frame,
                                                          bool robust) const {
  return findCheckerboardCorners(frame, pattern_size_, robust);
}

bool CameraCalibrator::addSample(const cv::Mat& frame) {
  auto corners = findCheckerboard(frame, true);
  if (!corners.has_value()) {
    return false;
  }

  const cv::Size size(frame.cols, frame.rows);
  if (!image_size_.has_value()) {
    image_size_ = size;
  } else if (*image_size_ != size) {
    std::ostringstream message;
    message << "All calibration frames must have the same resolution. Expected "
            << image_size_->width << "x" << image_size_->height << ", got " << size.width
            << "x" << size.height << ".";
    throw CalibrationError(message.str());
  }

  object_points_.push_back(object_template_.clone());
  image_points_.push_back(corners->clone());
  return true;
}

std::tuple<cv::Mat, cv::Mat, double> CameraCalibrator::calibrate() {
  if (sampleCount() < min_samples_) {
    std::ostringstream message;
    message << "Need at least " << min_samples_ << " valid samples, have " << sampleCount()
            << ".";
    throw CalibrationError(message.str());
  }
  if (!image_size_.has_value()) {
    throw CalibrationError("No calibration samples collected.");
  }

  std::vector<cv::Mat> rvecs;
  std::vector<cv::Mat> tvecs;
  const double error = cv::calibrateCamera(
      object_points_, image_points_, *image_size_, camera_matrix_, dist_coeffs_, rvecs, tvecs);
  reprojection_error_ = error;
  return {camera_matrix_.clone(), dist_coeffs_.clone(), error};
}

cv::Mat CameraCalibrator::drawOverlay(const cv::Mat& frame) const {
  cv::Mat overlay;
  auto corners = findCheckerboard(frame, false);
  if (corners.has_value()) {
    overlay = drawCheckerboardOverlay(frame, pattern_size_, *corners);
  } else {
    overlay = frame.clone();
  }

  const bool ready = sampleCount() >= min_samples_;
  std::ostringstream status;
  status << "Samples: " << sampleCount() << "/" << min_samples_ << "  "
         << (ready ? "READY" : "collect more");
  cv::putText(overlay,
              status.str(),
              cv::Point(20, 40),
              cv::FONT_HERSHEY_SIMPLEX,
              0.8,
              cv::Scalar(0, 255, 0),
              2,
              cv::LINE_AA);
  return overlay;
}

fs::path CameraCalibrator::saveToConfig(const std::optional<fs::path>& config_path) const {
  if (!isCalibrated() || !image_size_.has_value() || !reprojection_error_.has_value()) {
    throw CalibrationError("Calibrate before saving results.");
  }
  return saveCalibrationMatrices(camera_matrix_,
                                 dist_coeffs_,
                                 *reprojection_error_,
                                 *image_size_,
                                 config_path.value_or(config_path_));
}

void CameraCalibrator::resetSamples() {
  object_points_.clear();
  image_points_.clear();
  image_size_.reset();
}

}  // namespace chess_robot
